README.md
# NEX Switches Modules for VCV Rack

## Matrix Mixer 4

A matrix mixer with 4 inputs and 4 outputs. You can mute rows and/or columns via
some different algorithms.

*Un mélangeur matriciel avec 4 entrées et 4 sorties. Vous pouvez rendre muettes
les rangées et/ou les colonnes grâce à différents algorithmes.*

![](https://github.com/lkdjiin/SwitchesModules/blob/master/images/matrixmixer4.png)

**INPUTS**: The 4 inputs are on the top of the module.

**OUTPUTS**: The 4 outputs are on the right of the module.

**ROWS**: Think of each row as a 4x1 mixer. That is 4 inputs to 1 output.

**COLUMNS**: Think of each column as a 1x4 multiplexer. That is 1 input to 4 outputs with different levels.

**MUTE**: Pushing the button to the left of a row will mute/unmute that row.
Same thing for the columns, pushing the button at the bottom of a column will
mute/unmute that column. You can also use Control Voltage with each mute button.
How individual potentiometers will be enabled/disabled depends on the actual
algorithm.

**MUTE ALGORITHM**: There is 3 algorithms. The first one is «Force», the
second one is «Flip-flop», and the last one is «Intersections».
You can change the algorithm with the button on the bottom right.
You can also use Control Voltage to change the algorithm.

1. With «Force», the default one, the whole row (or
column) is enabled or disabled, regardless of the state of the individual
potentiometers.  That is, if a row (or column) is muted, all the 4
potentiometers of that row (or column) are disabled.  And reciprocally, if a
row (or column) is unmuted, all the 4 potentiometers of that row (or column)
are enabled. No doubt it's the most intuitive algorithm.
2. The «Flip-flop» works differently: when you mute (or
unmute) a row (or a column), each potentiometer's state is inversed in that row
(or column). That is, each enabled potentiometer of the row is disabled, and
each disabled potentiometer is enabled. I know this can be a confusing reading ;)
3. With «Intersections», potentiometers are enabled only if their respective row
and their respective column is enabled (or unmute, or on, or whatever you call it).

*NOTE: Despite the UI, there is only 3 algorithms for now, not 4.*

**AMPLITUDE ALGORITHM**: Right click the module to choose between ducking, hard clipping or no processing at all.
- Ducking - the more voices in a row, the less amplitude per voice.
- Hard clipping - 10 Volt peak to peak. This one could introduce a lot of distortion. Sometimes cool, sometimes harsh and not musical.
- No processing - Inputs are simply summed together.


## 2x2 Mute

It's like a crossfader for two groups of two channels. You can play group 1, or you can play group 2.
But you can't play both groups at the same time.

*C'est comme un fondu enchainé entre deux groupes de deux canaux. Vous jouez le groupe 1, ou vous jouez le groupe 2.
Mais vous ne pouvez pas jouer les deux groupes en même temps.*

![](https://github.com/lkdjiin/SwitchesModules/blob/master/images/2x2mute.png)

**IN**: On the picture above, the 2 inputs of group 1 are yellow and the 2 inputs of group 2 are red.

**OUT**: On the picture above, the 2 outputs of group 1 are green and the 2 outputs of group 2 are blue. First intput goes to first output, second input goes to second output and so on.

**Green LED Button**: Change the group to play. When the LED is on, group 1
 is playingplays and group 2 is muted. Vice versa, when the LED is off, group 1 is muted and group 2 is playing.

**CV**: Wait for a trigger then act like the green LED button.

**Knob**: Adjust the base time of the fade.

**Switch**: Adjust the scale for the fade's time, x1 up to 1 second, x10 up to 10 seconds and x100 up to 100 seconds.

**Exponential/Linear**: right click the module to choose between exponential and linear functions.

## Double Mute

With this module you can mute/unmute two monophonic inputs with a single button. You can
also choose the time of the fade-in and fade-out. Fade-in and fade-out are
independent and adjustable from 0 to 100 seconds.

*Avec ce module vous pouvez muter/démuter deux entrées (monophoniques) avec un seul bouton. Vous
pouvez aussi choisir le temps du fondu en ouverture et en fermeture. Ouverture
et fermeture sont indépendants et réglables de 0 à 100 secondes.*

![](https://github.com/lkdjiin/SwitchesModules/blob/master/images/doublemute.png)

**IN**: One or two input sources.

**OUT**: The outputs (top input goes to top output and bottom input goes to bottom output).

**Green LED Button**: Hit it to fade in/out.

**CV**: Wait for a trigger to fade in/out.

**Knobs**: Adjust the base time of the fade.

**Switches**: Adjust the scale for the fade's time, x1 up to 1 second, x10 up to 10 seconds and x100 up to 100 seconds.

**Exponential/Linear**: right click the module to choose between exponential and linear functions.


CHANGELOG.md
### 1.1.1 (2019-12-08)
- Add Matrix Mixer 4.

### 1.1.0 (2019-12-08)
*Never released*

### 1.0.0 (2019-11-06)
- Add Double Mute.
- Add 2 x 2 Mute.


src/DoubleMute.cpp
#include "plugin.hpp"

struct DoubleMute : Module {

    const float PRESET_FADE = 0.1;

    enum ParamIds {
        MUTE_PARAM,
        FADE_IN_PARAM,
        FADE_OUT_PARAM,
        SCALE_IN_PARAM,
        SCALE_OUT_PARAM,
        NUM_PARAMS
    };
    enum InputIds {
        IN1_INPUT,
        IN2_INPUT,
        TRIGGER_MUTE_INPUT,
        NUM_INPUTS
    };
    enum OutputIds {
        OUT1_OUTPUT,
        OUT2_OUTPUT,
        NUM_OUTPUTS
    };
    enum LightIds {
        MUTE_LIGHT,
        RAMP_UP_LIGHT,
        RAMP_DOWN_LIGHT,
        NUM_LIGHTS
    };

    enum States {
        HIGH,
        LOW,
        RAMP_UP,
        RAMP_DOWN
    };

    States state;
    dsp::BooleanTrigger muteTrigger;
    float fadeTimeEllapsed;
    bool exponentialFade = true; // If not it's linear.

    DoubleMute() {
        config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        configParam(MUTE_PARAM, 0.0, 1.0, 0.0, "Mute channels");
        configParam(FADE_IN_PARAM, 0.01, 1.0, PRESET_FADE, "Fade-in time");
        configParam(FADE_OUT_PARAM, 0.01, 1.0, PRESET_FADE, "Fade-out time");
        configParam(SCALE_IN_PARAM, 0.0, 2.0, 0.0, "Fade-in scale");
        configParam(SCALE_OUT_PARAM, 0.0, 2.0, 0.0, "Fade-out scale");

        onReset();
    }

    void onReset() override {
        lights[RAMP_UP_LIGHT].setBrightness(0.f);
        lights[RAMP_DOWN_LIGHT].setBrightness(0.f);
    }

    void process(const ProcessArgs& args) override {
        setState();

        switch(state) {
            case HIGH:
                lights[MUTE_LIGHT].setBrightness(0.9f);
                high();
                break;
            case RAMP_UP:
                rampUp(args.sampleTime);
                break;
            case RAMP_DOWN:
                rampDown(args.sampleTime);
                break;
            case LOW:
            default:
                break;
        }
    }

    void setState() {
        if (muteTrigger.process(params[MUTE_PARAM].getValue() > 0.f)
                || inputs[TRIGGER_MUTE_INPUT].getNormalVoltage(0.f) > 0.f) {
            switch(state) {
                case HIGH:
                    state = RAMP_DOWN;
                    fadeTimeEllapsed = rampDownTime();
                    lights[MUTE_LIGHT].setBrightness(0.f);
                    lights[RAMP_DOWN_LIGHT].setBrightness(0.9f);
                    lights[RAMP_UP_LIGHT].setBrightness(0.f);
                    break;
                case RAMP_UP:
                    state = RAMP_DOWN;
                    fadeTimeEllapsed = rampUpToDownTime();
                    lights[MUTE_LIGHT].setBrightness(0.f);
                    lights[RAMP_DOWN_LIGHT].setBrightness(0.9f);
                    lights[RAMP_UP_LIGHT].setBrightness(0.f);
                    break;
                case RAMP_DOWN:
                    state = RAMP_UP;
                    fadeTimeEllapsed = rampDownToUpTime();
                    lights[MUTE_LIGHT].setBrightness(0.9f);
                    lights[RAMP_DOWN_LIGHT].setBrightness(0.f);
                    lights[RAMP_UP_LIGHT].setBrightness(0.9f);
                    break;
                case LOW:
                default:
                    state = RAMP_UP;
                    fadeTimeEllapsed = 0.f;
                    lights[MUTE_LIGHT].setBrightness(0.9f);
                    lights[RAMP_DOWN_LIGHT].setBrightness(0.f);
                    lights[RAMP_UP_LIGHT].setBrightness(0.9f);
                    break;
            }
        }
    }

    float rampUpToDownTime() {
        float ratio = fadeTimeEllapsed / rampUpTime();
        return rampDownTime() * ratio;
    }

    float rampDownToUpTime() {
        float ratio = fadeTimeEllapsed / rampDownTime();
        return rampUpTime() * ratio;
    }

    float rampDownTime() {
        float time = params[FADE_OUT_PARAM].getValue();
        int mult = (int) std::round(params[SCALE_OUT_PARAM].getValue());
        switch(mult) {
            case 1:
                time *= 10;
                break;
            case 2:
                time *= 100;
                break;
            default:
                break;
        }
        return time;
    }

    float rampUpTime() {
        float time = params[FADE_IN_PARAM].getValue();
        int mult = (int) std::round(params[SCALE_IN_PARAM].getValue());
        switch(mult) {
            case 1:
                time *= 10;
                break;
            case 2:
                time *= 100;
                break;
            default:
                break;
        }
        return time;
    }

    void rampUp(float sampleTime) {
        fadeTimeEllapsed += sampleTime;
        float userValue = rampUpTime();
        float mult = fadeTimeEllapsed / userValue;
        mult = clamp(mult, 0.f, 1.f);

        if (exponentialFade) {
            mult = rescale(std::pow(50.f, mult), 1.f, 50.f, 0.f, 1.f);
        }

        if (inputs[IN1_INPUT].isConnected() && outputs[OUT1_OUTPUT].isConnected()) {
            outputs[OUT1_OUTPUT].setVoltage(inputs[IN1_INPUT].getVoltage() * mult);
        }
        if (inputs[IN2_INPUT].isConnected() && outputs[OUT2_OUTPUT].isConnected()) {
            outputs[OUT2_OUTPUT].setVoltage(inputs[IN2_INPUT].getVoltage() * mult);
        }
        if (fadeTimeEllapsed >= userValue) {
            state = HIGH;
            lights[RAMP_UP_LIGHT].setBrightness(0.f);
        } else {
            lights[RAMP_UP_LIGHT].setBrightness(1.f - (mult));
        }
    }

    void high() {
        if (inputs[IN1_INPUT].isConnected() && outputs[OUT1_OUTPUT].isConnected()) {
            outputs[OUT1_OUTPUT].setVoltage(inputs[IN1_INPUT].getVoltage());
        }
        if (inputs[IN2_INPUT].isConnected() && outputs[OUT2_OUTPUT].isConnected()) {
            outputs[OUT2_OUTPUT].setVoltage(inputs[IN2_INPUT].getVoltage());
        }
    }

    void rampDown(float sampleTime) {
        fadeTimeEllapsed -= sampleTime;
        float userValue = rampDownTime();
        float mult = fadeTimeEllapsed / userValue;
        mult = clamp(mult, 0.f, 1.f);

        if (exponentialFade) {
            mult = rescale(std::pow(50.f, mult), 1.f, 50.f, 0.f, 1.f);
        }

        if (inputs[IN1_INPUT].isConnected() && outputs[OUT1_OUTPUT].isConnected()) {
            outputs[OUT1_OUTPUT].setVoltage(inputs[IN1_INPUT].getVoltage() * mult);
        }
        if (inputs[IN2_INPUT].isConnected() && outputs[OUT2_OUTPUT].isConnected()) {
            outputs[OUT2_OUTPUT].setVoltage(inputs[IN2_INPUT].getVoltage() * mult);
        }
        if (fadeTimeEllapsed <= 0.f) {
            state = LOW;
            lights[RAMP_DOWN_LIGHT].setBrightness(0.f);
        } else {
            lights[RAMP_DOWN_LIGHT].setBrightness(mult);
        }
    }

    json_t *dataToJson() override {
        float value;
        value = (state == HIGH || state == RAMP_UP) ? 1 : 0;
        json_t *rootJ = json_object();
        json_object_set_new(rootJ, "state", json_integer(value));
        json_object_set_new(rootJ, "exponentialFade", json_integer((int) exponentialFade));
        return rootJ;
    }

    void dataFromJson(json_t *rootJ) override {
        json_t *stateJ = json_object_get(rootJ, "state");
        if (stateJ) {
            float value = json_integer_value(stateJ);
            if (value == 1) {
                state = HIGH;
            } else {
                state = LOW;
            }
        }

        json_t *exponentialFadeJ = json_object_get(rootJ, "exponentialFade");
        if (exponentialFadeJ) {
            exponentialFade = json_integer_value(exponentialFadeJ);
        }
    }
};

template <typename BASE>
struct MuteLight : BASE {
    MuteLight() {
        this->box.size = mm2px(Vec(6.f, 6.f));
    }
};

struct DoubleMuteWidget : ModuleWidget {
    DoubleMuteWidget(DoubleMute* module) {
        setModule(module);
        setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/DoubleMute.svg")));

        addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH,
                        RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
        addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH,
                        RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

        addParam(createParamCentered<LEDBezel>(mm2px(Vec(15.24, 32.0)),
                    module, DoubleMute::MUTE_PARAM));
        addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(15.24, 68.5)),
                    module, DoubleMute::FADE_IN_PARAM));
        addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(15.24, 99.5)),
                    module, DoubleMute::FADE_OUT_PARAM));
        addParam(createParamCentered<CKSSThree>(mm2px(Vec(15.24, 81.0)),
                    module, DoubleMute::SCALE_IN_PARAM));
        addParam(createParamCentered<CKSSThree>(mm2px(Vec(15.24, 112.0)),
                    module, DoubleMute::SCALE_OUT_PARAM));

        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 22.0)),
                    module, DoubleMute::IN1_INPUT));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 42.0)),
                    module, DoubleMute::IN2_INPUT));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(15.24, 52.0)),
                    module, DoubleMute::TRIGGER_MUTE_INPUT));

        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 22.0)),
                    module, DoubleMute::OUT1_OUTPUT));
        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 42.0)),
                    module, DoubleMute::OUT2_OUTPUT));

        addChild(createLightCentered<MuteLight<GreenLight>>(mm2px(Vec(15.24, 32.0)),
                    module, DoubleMute::MUTE_LIGHT));
        addChild(createLightCentered<SmallLight<YellowLight>>(mm2px(Vec(22.86, 68.5)),
                    module, DoubleMute::RAMP_UP_LIGHT));
        addChild(createLightCentered<SmallLight<YellowLight>>(mm2px(Vec(22.86, 99.5)),
                    module, DoubleMute::RAMP_DOWN_LIGHT));
    }

    struct DoubleMuteModeItem : MenuItem {
        DoubleMute *module;
        bool mode;
        void onAction(const event::Action &e) override {
            module->exponentialFade = mode;
        }
        void step() override {
            rightText = (module->exponentialFade == mode) ? "✔" : "";
        }
    };

    void appendContextMenu(Menu *menu) override {
        MenuLabel *spacerLabel = new MenuLabel();
        menu->addChild(spacerLabel);

        DoubleMute *module = dynamic_cast<DoubleMute*>(this->module);
        assert(module);

        MenuLabel *themeLabel = new MenuLabel();
        themeLabel->text = "Audio Mode";
        menu->addChild(themeLabel);

        DoubleMuteModeItem *item1 = new DoubleMuteModeItem();
        item1->text = "Exponential";
        item1->module = module;
        item1->mode = true;
        menu->addChild(item1);

        DoubleMuteModeItem *item2 = new DoubleMuteModeItem();
        item2->text = "Linear";
        item2->module = module;
        item2->mode = false;
        menu->addChild(item2);
    }

};


Model* modelDoubleMute = createModel<DoubleMute, DoubleMuteWidget>("DoubleMute");


src/TwoByTwo.cpp
#include "plugin.hpp"

struct TwoByTwo : Module {

    const float PRESET_FADE = 0.1;

    enum ParamIds {
        MUTE_PARAM,
        FADE_PARAM,
        SCALE_PARAM,
        NUM_PARAMS
    };

    enum InputIds {
        ENUMS(IN_INPUTS, 4),
        TRIGGER_MUTE_INPUT,
        NUM_INPUTS
    };

    enum OutputIds {
        ENUMS(OUT_OUTPUTS, 4),
        NUM_OUTPUTS
    };

    enum LightIds {
        MUTE_LIGHT,
        RAMP_LIGHT,
        GROUP1_LIGHT,
        GROUP2_LIGHT,
        NUM_LIGHTS
    };

    enum States {
        HIGH,
        LOW,
        RAMP_UP,
        RAMP_DOWN
    };

    States state;
    dsp::BooleanTrigger muteTrigger;
    float fadeInTime;
    float fadeOutTime;
    bool exponentialFade = false; // If not it's linear.

    TwoByTwo() {
        config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
        configParam(MUTE_PARAM, 0.0, 1.0, 0.0, "Mute channels");
        configParam(FADE_PARAM, 0.01, 1.0, PRESET_FADE, "Fade-in/out time");
        configParam(SCALE_PARAM, 0.0, 2.0, 0.0, "Fade-in/out scale");

        onReset();
    }

    void onReset() override {
        lights[RAMP_LIGHT].setBrightness(0.f);
        lights[GROUP1_LIGHT].setBrightness(0.f);
        lights[GROUP2_LIGHT].setBrightness(0.f);
    }

    // We need an initial state. For that matter, HIGH is as good as LOW.
    void onAdd() override {
        if (state != HIGH && state != LOW) {
            state = HIGH;
        }
    }

    void process(const ProcessArgs& args) override {
        setState();

        switch(state) {
            case HIGH:
                high();
                break;
            case RAMP_UP:
                // Fade in group 1.
                rampUp(args.sampleTime, 0, 1);
                // Fade out group 2.
                rampDown(args.sampleTime, 2, 3);
                break;
            case RAMP_DOWN:
                // Fade out group 1.
                rampDown(args.sampleTime, 0, 1);
                // Fade in group 2.
                rampUp(args.sampleTime, 2, 3);
                break;
            case LOW:
                low();
                break;
            default:
                printf("*** NO STATE :(((\n");
                break;
        }
    }

    void setLights(float muteL, float rampL, float group1L, float group2L) {
        lights[MUTE_LIGHT].setBrightness(muteL);
        lights[RAMP_LIGHT].setBrightness(rampL);
        lights[GROUP1_LIGHT].setBrightness(group1L);
        lights[GROUP2_LIGHT].setBrightness(group2L);
    }

    void setState() {
        if (muteTrigger.process(params[MUTE_PARAM].getValue() > 0.f)
                || inputs[TRIGGER_MUTE_INPUT].getNormalVoltage(0.f) > 0.f) {
            switch(state) {
                case HIGH:
                    state = RAMP_DOWN;
                    fadeOutTime = rampTime();
                    fadeInTime = 0.f;
                    setLights(0.f, 0.9f, 0.f, 0.9f);
                    break;
                case RAMP_UP:
                    state = RAMP_DOWN;
                    exchangeFades();
                    setLights(0.f, 0.9f, 0.f, 0.9f);
                    break;
                case RAMP_DOWN:
                    state = RAMP_UP;
                    exchangeFades();
                    setLights(0.9f, 0.9f, 0.9f, 0.f);
                    break;
                case LOW:
                    state = RAMP_UP;
                    fadeOutTime = rampTime();
                    fadeInTime = 0.f;
                    setLights(0.9f, 0.9f, 0.9f, 0.f);
                    break;
                default:
                    printf("*** NO STATE WHEN TRIGGERED!\n");
                    break;
            }
        }
    }

    // This is useful when we trigger a RAMP_UP from a RAMP_DOWN state
    // and vice versa. We exchange fade-in and fade-out to (kind of)
    // move backward.
    void exchangeFades() {
        float temp = fadeInTime;
        fadeInTime = fadeOutTime;
        fadeOutTime = temp;
    }

    // The time of the fade-in/out, as set by the user.
    float rampTime() {
        float time = params[FADE_PARAM].getValue();
        int mult = (int) std::round(params[SCALE_PARAM].getValue());
        switch(mult) {
            case 1:
                time *= 10;
                break;
            case 2:
                time *= 100;
                break;
            default:
                break;
        }
        return time;
    }

    // Play group 1, full amplitude.
    void high() {
        playFullInput(0);
        playFullInput(1);
    }

    // Play group 2, full amplitude.
    void low() {
        playFullInput(2);
        playFullInput(3);
    }

    inline void playFullInput(int inputNumber) {
        if (inputs[IN_INPUTS + inputNumber].isConnected() &&
               outputs[OUT_OUTPUTS + inputNumber].isConnected()) {
            outputs[OUT_OUTPUTS + inputNumber].setVoltage(
                    inputs[IN_INPUTS + inputNumber].getVoltage());
        }
    }

    inline void playPartInput(int inputNumber, float mult) {
        if (inputs[IN_INPUTS + inputNumber].isConnected() &&
               outputs[OUT_OUTPUTS + inputNumber].isConnected()) {
            outputs[OUT_OUTPUTS + inputNumber].setVoltage(
                    inputs[IN_INPUTS + inputNumber].getVoltage() * mult);
        }
    }

    // Fade-in the 2 inputs of a given group.
    void rampUp(float sampleTime, int channelA, int channelB) {
        fadeInTime += sampleTime;
        float userValue = rampTime();
        float mult = fadeInTime / userValue;
        mult = clamp(mult, 0.f, 1.f);

        if (exponentialFade) {
            mult = rescale(std::pow(50.f, mult), 1.f, 50.f, 0.f, 1.f);
        }

        playPartInput(channelA, mult);
        playPartInput(channelB, mult);

        if (fadeInTime >= userValue) {
            if (state == RAMP_UP) {
                state = HIGH;
                lights[RAMP_LIGHT].setBrightness(0.f);
            }
        }
    }

    // Fade-out the 2 inputs of a given group.
    void rampDown(float sampleTime, int channelA, int channelB) {
        fadeOutTime -= sampleTime;
        float userValue = rampTime();
        float mult = fadeOutTime / userValue;
        mult = clamp(mult, 0.f, 1.f);

        if (exponentialFade) {
            mult = rescale(std::pow(50.f, mult), 1.f, 50.f, 0.f, 1.f);
        }

        playPartInput(channelA, mult);
        playPartInput(channelB, mult);

        if (fadeOutTime <= 0.f) {
            if (state == RAMP_DOWN) {
                state = LOW;
                lights[RAMP_LIGHT].setBrightness(0.f);
            }
        } else {
            lights[RAMP_LIGHT].setBrightness(mult);
        }
    }

    json_t *dataToJson() override {
        float value;
        value = (state == HIGH || state == RAMP_UP) ? 1 : 0;
        json_t *rootJ = json_object();
        json_object_set_new(rootJ, "state", json_integer(value));
        json_object_set_new(rootJ, "exponentialFade", json_integer((int) exponentialFade));
        return rootJ;
    }

    void dataFromJson(json_t *rootJ) override {
        json_t *stateJ = json_object_get(rootJ, "state");
        if (stateJ) {
            float value = json_integer_value(stateJ);
            if (value == 1) {
                state = HIGH;
                setLights(0.9f, 0.f, 0.9f, 0.f);
            } else {
                state = LOW;
                setLights(0.f, 0.f, 0.f, 0.9f);
            }
        }

        json_t *exponentialFadeJ = json_object_get(rootJ, "exponentialFade");
        if (exponentialFadeJ) {
            exponentialFade = json_integer_value(exponentialFadeJ);
        }
    }
};

template <typename BASE>
struct MuteLight : BASE {
    MuteLight() {
        this->box.size = mm2px(Vec(6.f, 6.f));
    }
};

struct TwoByTwoWidget : ModuleWidget {
    TwoByTwoWidget(TwoByTwo* module) {
        setModule(module);
        setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/2x2Mute.svg")));

        addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
        addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH,
                        RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
        addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH,
                        RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

        addParam(createParamCentered<LEDBezel>(mm2px(Vec(15.24, 42.0)),
                    module, TwoByTwo::MUTE_PARAM));
        addParam(createParamCentered<RoundBlackKnob>(mm2px(Vec(15.24, 99.5)),
                    module, TwoByTwo::FADE_PARAM));
        addParam(createParamCentered<CKSSThree>(mm2px(Vec(15.24, 112.0)),
                    module, TwoByTwo::SCALE_PARAM));

        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 22.0)),
                    module, TwoByTwo::IN_INPUTS + 0));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 32.0)),
                    module, TwoByTwo::IN_INPUTS + 1));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 62.0)),
                    module, TwoByTwo::IN_INPUTS + 2));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.62, 72.0)),
                    module, TwoByTwo::IN_INPUTS + 3));
        addInput(createInputCentered<PJ301MPort>(mm2px(Vec(15.24, 52.0)),
                    module, TwoByTwo::TRIGGER_MUTE_INPUT));

        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 22.0)),
                    module, TwoByTwo::OUT_OUTPUTS + 0));
        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 32.0)),
                    module, TwoByTwo::OUT_OUTPUTS + 1));
        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 62.0)),
                    module, TwoByTwo::OUT_OUTPUTS + 2));
        addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(22.86, 72.0)),
                    module, TwoByTwo::OUT_OUTPUTS + 3));

        addChild(createLightCentered<MuteLight<GreenLight>>(mm2px(Vec(15.24, 42.0)),
                    module, TwoByTwo::MUTE_LIGHT));
        addChild(createLightCentered<SmallLight<YellowLight>>(mm2px(Vec(15.24, 91.5)),
                    module, TwoByTwo::RAMP_LIGHT));
        addChild(createLightCentered<SmallLight<WhiteLight>>(mm2px(Vec(15.24, 27.0)),
                    module, TwoByTwo::GROUP1_LIGHT));
        addChild(createLightCentered<SmallLight<WhiteLight>>(mm2px(Vec(15.24, 67.0)),
                    module, TwoByTwo::GROUP2_LIGHT));
    }

    struct TwoByTwoModeItem : MenuItem {
        TwoByTwo *module;
        bool mode;
        void onAction(const event::Action &e) override {
            module->exponentialFade = mode;
        }
        void step() override {
            rightText = (module->exponentialFade == mode) ? "✔" : "";
        }
    };

    void appendContextMenu(Menu *menu) override {
        MenuLabel *spacerLabel = new MenuLabel();
        menu->addChild(spacerLabel);

        TwoByTwo *module = dynamic_cast<TwoByTwo*>(this->module);
        assert(module);

        MenuLabel *themeLabel = new MenuLabel();
        themeLabel->text = "Audio Mode";
        menu->addChild(themeLabel);

        TwoByTwoModeItem *item1 = new TwoByTwoModeItem();
        item1->text = "Exponential";
        item1->module = module;
        item1->mode = true;
        menu->addChild(item1);

        TwoByTwoModeItem *item2 = new TwoByTwoModeItem();
        item2->text = "Linear";
        item2->module = module;
        item2->mode = false;
        menu->addChild(item2);
    }

};

Model* modelTwoByTwo = createModel<TwoByTwo, TwoByTwoWidget>("TwoByTwo");
